### PR TITLE
Fixed issue with model vertex color in rgb format

### LINKED ISSUE #7366: Random colors from a GLB file are displayed
**Describe the bug (REQUIRED)**
It turned out that Defold can import colors from glTF 3D models. But the feature doesn't work.

image

Plus, Defold forces to specify a texture for a model to bundle the project. Why?
image

**To Reproduce (REQUIRED)**
Take a look at the included project.

**Expected behavior (REQUIRED)**
The 3d model should be colored.

**Defold version (REQUIRED):**
  - Version: 1.4.2

**Platforms (REQUIRED):**
 - Platforms: Windows 10

**Minimal repro case project (OPTIONAL):**
[issue_glb_colors.zip](https://github.com/defold/defold/files/10611393/issue_glb_colors.zip)

The GLB file was exported from Houdini Indie 19.5.

**Screenshots (OPTIONAL):**
The included GLB file is valid (https://github.khronos.org/glTF-Validator/):
image

glb.ee and other online glTF viewers display the model perfectly fine
image

### PR DESCRIPTION
**Pull request by JCash**
This fixes an issue when importing vertex colors in 3-tuples. We convert it into 4-tuples, with `alpha=1.0f`.
It is now also possible to avoid setting any textures at all in case your model doesn't need them.

Fixes #7366 

## PR checklist

* [x] Code
	* [x] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [x] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [x] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [x] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelBuilder.java [hunk 1]
```diff
@@ -60,6 +60,9 @@ public Task<Void> create(IResource input) throws IOException, CompileExceptionEr
             taskBuilder.addInput(animations);
         }
         for (String t : modelDescBuilder.getTexturesList()) {
+            if (t.isEmpty())
+                continue; // TODO: Perhaps we can check if the material expects textures?
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelBuilder.java:64]: It wasn't possible to build a model with no textures.

File: engine/modelc/src/modelimporter_gltf.cpp [hunk 2]
```diff
@@ -149,16 +149,31 @@ static const char* GetInterpolationTypeStr(cgltf_interpolation_type type)
...
 {
     uint32_t num_components = (uint32_t)cgltf_num_components(accessor->type);
 
     if (desired_num_components == 0)
         desired_num_components = num_components;
 
-    float* out = new float[accessor->count * num_components];
+    uint32_t size = accessor->count * num_components;
+    if (desired_num_components > num_components)
+        size = accessor->count * desired_num_components;
```
Comment (JCash) [inline engine/modelc/src/modelimporter_gltf.cpp:161]: The buffer might have 3 components, but we want 4.

File: engine/modelc/src/modelimporter_gltf.cpp [hunk 3]
```diff
@@ -149,16 +149,31 @@ static const char* GetInterpolationTypeStr(cgltf_interpolation_type type)
...
+    float* out = new float[size]; // Now the buffer will fit the max num components
     float* writeptr = out;
 
+    if (desired_num_components > num_components)
+    {
+        for (uint32_t i = 0; i < accessor->count; ++i)
+        {
+            for (uint32_t j = 0; j < desired_num_components; ++j)
+                *writeptr++ = default_value;
+        }
```
Comment (JCash) [inline engine/modelc/src/modelimporter_gltf.cpp:172]: I couldn't figure out a better memset for floats 🤔

File: engine/modelc/src/modelimporter_gltf.cpp [hunk 4]
```diff
@@ -412,17 +427,24 @@ static void LoadPrimitives(Model* model, cgltf_data* gltf_data, cgltf_mesh* gltf
...
+
+            float default_value_f = 0.0f;
             if (attribute->type == cgltf_attribute_type_tangent)
             {
                 desired_num_components = 3; // for some reason it give 4 elements
             }
+            else if (attribute->type == cgltf_attribute_type_color)
+            {
+                desired_num_components = 4; // We currently always store vec4
+                default_value_f = 1.0f; // for the alpha channel
```
Comment (JCash) [inline engine/modelc/src/modelimporter_gltf.cpp:446]: Perhaps in the future we'll support only rgb colors, but for now this is the quickest way forward.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelBuilder.java (hunk 1 above)
Comment (Jhonnyg) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelBuilder.java:64]: We can check if there's samplers in the material shaders, but I don't think it should be a build error. In the case of the PBR shaders for instance, we want to be able to set global textures for all models in runtime

File: engine/modelc/src/test/test_model.cpp [hunk 5]
```diff
@@ -93,6 +93,27 @@ TEST(ModelGLTF, LoadSkeleton)
...
 
+// See https://github.com/KhronosGroup/glTF-Asset-Generator for assets
+// https://raw.githubusercontent.com/KhronosGroup/glTF-Sample-Models/master/2.0/VertexColorTest/glTF-Embedded/VertexColorTest.gltf
+
+TEST(ModelGLTF, VertexColor3Float)
+{
+    // Model courtesy of Artsion Trubchyk (https://github.com/aglitchman), public domain
+    const char* path = "./src/test/assets/primitive_vertex_color/vertexcolor_rgb3.gltf";
+    uint32_t file_size = 0;
+    void* mem = dmModelImporter::ReadFile(path, &file_size);
```
Comment (Jhonnyg) [inline engine/modelc/src/test/test_model.cpp:104]: Just a thought, but don't we usually return a result enum and pass in pointers to the functions? 🤔 

Edit: Ah, it's only used for tests..

**Review by Jhonnyg — APPROVED** [review]
👍

**Pull request merged**